### PR TITLE
Add `withExceptions` attribute to `@ExpectedToFail`

### PR DESCRIPTION
**Pull request by knutwannheden**
Adds a new `onExceptions` attribute to `@ExpectedToFail` using which the scope of the extension can be limited to only consider the test as successful if the thrown exception matches one of the given types.

Fixes: #769

### CODE REVIEW

**Comment by knutwannheden** [comment]
This is still in draft state, as I will need to add more documentation and tests. I just wanted to make sure I am not pulling in the wrong direction with this and also check if you already have some input on this.

File: src/main/java/org/junitpioneer/jupiter/ExpectedToFailExtension.java [hunk 1]
```diff
@@ -42,7 +42,12 @@ private static void invokeAndInvertResult(Invocation<Void> invocation, Extension
 				throw t;
 			}
 
-			String message = getExpectedToFailAnnotation(extensionContext).value();
+			ExpectedToFail expectedToFail = getExpectedToFailAnnotation(extensionContext);
+			if (!expectedToFail.onExceptions()[0].isInstance(t)) {
```
Comment (Michael1993) [inline src/main/java/org/junitpioneer/jupiter/ExpectedToFailExtension.java:46]: I thought this was an array but you only check the first element?

File: src/test/java/org/junitpioneer/jupiter/ExpectedToFailExtensionTests.java [hunk 2]
```diff
@@ -101,6 +101,16 @@ void failsOnWorkingTest() {
...
 	}
 
+	@Test
+	void failsOnTestThrowingUnexpectedException() {
+		ExecutionResults results = PioneerTestKit.executeTestMethod(ExpectedToFailTestCases.class, "unexpectedException");
+		assertThat(results)
+				.hasSingleStartedTest()
+				.whichFailed()
+				.withExceptionInstanceOf(AssertionError.class)
+				.hasMessage("Test marked as 'expected to fail' failed with an unexpected class java.lang.IllegalArgumentException exception");
```
Comment (Michael1993) [inline src/test/java/org/junitpioneer/jupiter/ExpectedToFailExtensionTests.java:111]: I would not bake in the classname to the test, but that's personal preference.

The exception is actually an [AssertionFailedError](https://ota4j-team.github.io/opentest4j/docs/1.0.0/api/org/opentest4j/AssertionFailedError.html?is-external=true) an extension of but not an `AssertionError` (exactly).

File: src/main/java/org/junitpioneer/jupiter/ExpectedToFailExtension.java [hunk 3]
```diff
@@ -42,7 +42,12 @@ private static void invokeAndInvertResult(Invocation<Void> invocation, Extension
...
+				fail("Test marked as 'expected to fail' failed with an unexpected " + t.getClass() + " exception");
```
Comment (Michael1993) [inline src/main/java/org/junitpioneer/jupiter/ExpectedToFailExtension.java:47]: `Assertions::fail` can take a second parameter `Throwable` cause, which I think is applicable here?

**Review by Michael1993 — COMMENTED** [review]
I am undecided if I like this proposal. I really appreciate you taking the initiative and I'm not outright opposed to it, it's just that I worry that despite this being something for a specific use-case, people will be tempted to use it as a "fix-all". Does that make sense?

**Comment by knutwannheden** [comment]
@Michael1993 I definitely agree with your review comments. I just slapped this together to better convey the idea and to have something concrete to discuss.

I understand your concern but I honestly don't quite see the "fix-all" problem. If anything it is JUnit's own `@Disabled` I see as a fix-all, that gets used (and then forgotten!) far too often. Also `@ExpectedToFail` annotations can linger on for a long time, which I think is OK, since they already narrow down the scope a lot (the test actually must fail). Now, this `onExceptions` attribute narrows this scope down even more, because the test must fail with a specific exception.

My only "concern" as a maintainer would be that the use case is probably rather unusual. But as I think I already mentioned in the issue, JUnit 4's `@Test(expected)` actually allowed this (even though it was probably primarily designed for what today `assertThrows()` should be used for) and I found it extremely useful when for example working on a new implementation for some API spec.

**Comment by Michael1993** [comment]
Okay, I see your point. Please go ahead with the PR. 😄

File: src/main/java/org/junitpioneer/jupiter/ExpectedToFailExtension.java (hunk 3 above)
Comment (knutwannheden) [reply src/main/java/org/junitpioneer/jupiter/ExpectedToFailExtension.java:47]: So you mean that if the test throws an `AssertionFailedError` then we should check its cause (if any) against the `onExceptions` types?

Comment (Michael1993) [reply src/main/java/org/junitpioneer/jupiter/ExpectedToFailExtension.java]: Hm, no, I don't think that's right. Assertion errors are assertion failures which is not the use-case I had in mind... Unless you can justify it?

I just meant we should not be "swallowing" unexpected exceptions because it makes debugging harder.

File: src/main/java/org/junitpioneer/jupiter/ExpectedToFail.java [hunk 4]
```diff
@@ -33,6 +33,13 @@
  * of disabled tests over time.
  * </p>
  *
+ * <p>Further, the {@link #onExceptions()} attribute can be used to restrict the extension's behavior
+ * to specific exceptions. I.e. only if the test method end up throwing one of the specified exceptions
+ * will the test be aborted. This can for example be used when the production code temporarily throws
+ * an {@link UnsupportedOperationException} because some feature has not been implemented yet, but the
+ * test method is already implemented and should not be aborted on a failing assertion.
```
Comment (Michael1993) [inline src/main/java/org/junitpioneer/jupiter/ExpectedToFail.java:40]: Should not be aborted?

This doesn't really makes sense. I think this should be "should not fail"?

**Review by Michael1993 — CHANGES_REQUESTED** [review]
Please apply spotless and add documentation, if you can. Thank you!

File: src/main/java/org/junitpioneer/jupiter/ExpectedToFailExtension.java (hunk 3 above)
Comment (knutwannheden) [reply src/main/java/org/junitpioneer/jupiter/ExpectedToFailExtension.java]: OK, I see what you mean. That of course makes sense.

File: src/demo/java/org/junitpioneer/jupiter/ExpectedToFailExtensionDemo.java [hunk 5]
```diff
@@ -18,7 +18,7 @@ public class ExpectedToFailExtensionDemo {
 
 	// tag::expected_to_fail[]
 	@Test
-	@ExpectedToFail
+	@ExpectedToFail(onExceptions = Exception.class)
```
Comment (Michael1993) [inline src/demo/java/org/junitpioneer/jupiter/ExpectedToFailExtensionDemo.java:21]: Sorry, why did you change this?

File: src/main/java/org/junitpioneer/jupiter/ExpectedToFailExtension.java [hunk 6]
```diff
@@ -42,7 +43,12 @@ private static void invokeAndInvertResult(Invocation<Void> invocation, Extension
 				throw t;
 			}
 
-			String message = getExpectedToFailAnnotation(extensionContext).value();
+			ExpectedToFail expectedToFail = getExpectedToFailAnnotation(extensionContext);
+			if (Stream.of(expectedToFail.onExceptions()).noneMatch(c -> c.isInstance(t))) {
```
Comment (Michael1993) [inline src/main/java/org/junitpioneer/jupiter/ExpectedToFailExtension.java:47]: Personally, I am not a fan of single letter lambda parameters but that is just personal preference. 🐸

File: src/demo/java/org/junitpioneer/jupiter/ExpectedToFailExtensionDemo.java (hunk 5 above)
Comment (knutwannheden) [reply src/demo/java/org/junitpioneer/jupiter/ExpectedToFailExtensionDemo.java:21]: Hmm... That looks like a mistake. I think I probably did that when I first started looking into this and mistakenly edited this file. Sorry about that!

File: src/main/java/org/junitpioneer/jupiter/ExpectedToFailExtension.java (hunk 6 above)
Comment (knutwannheden) [reply src/main/java/org/junitpioneer/jupiter/ExpectedToFailExtension.java]: For me it is either way. Paramount is of course the overall code consistency.

File: src/demo/java/org/junitpioneer/jupiter/ExpectedToFailExtensionDemo.java [hunk 7]
```diff
@@ -32,10 +32,23 @@ void doSomething() {
 		int actual = brokenMethod();
 		assertEquals(10, actual);
 	}
-	// end::expected_to_fail_message[]
 
 	private int brokenMethod() {
 		return 0;
 	}
+	// end::expected_to_fail_message[]
```
Comment (knutwannheden) [inline src/demo/java/org/junitpioneer/jupiter/ExpectedToFailExtensionDemo.java:39]: I hope you agree with me moving this end marker and doing the same for my new example.

Comment (Michael1993) [reply src/demo/java/org/junitpioneer/jupiter/ExpectedToFailExtensionDemo.java:39]: I don't think the two are equivalent. Your example showcases a planned feature with no body (which could be an actual example of someone using this) but the previous example doesn't actually showcase what the example is talking about... The `brokenMethod` is not actually "broken", and seeing it **not** returning 10 doesn't help understanding. Does this make sense?

I'd move back the tag here but leave the one in your example where it is.

File: docs/expected-to-fail-tests.adoc [hunk 8]
```diff
@@ -44,6 +44,18 @@ A custom message can be provided, explaining why the tested code is not working
 include::{demo}[tag=expected_to_fail_message]
 ----
 
+In some cases it may be desirable to only have the test be aborted on certain exceptions.
```
Comment (Michael1993) [inline docs/expected-to-fail-tests.adoc:47]: I like the documentation change but the project adheres to the "one-sentence-per-line" rule. Could you format it that way, please?

File: src/demo/java/org/junitpioneer/jupiter/ExpectedToFailExtensionDemo.java (hunk 7 above)
Comment (knutwannheden) [reply src/demo/java/org/junitpioneer/jupiter/ExpectedToFailExtensionDemo.java:39]: Yes, that makes sense.

File: docs/expected-to-fail-tests.adoc (hunk 8 above)
Comment (knutwannheden) [reply docs/expected-to-fail-tests.adoc:47]: I should have read that more carefully. Sorry about that!

File: docs/expected-to-fail-tests.adoc [hunk 9]
```diff
@@ -45,11 +45,9 @@ include::{demo}[tag=expected_to_fail_message]
...
 
 In some cases it may be desirable to only have the test be aborted on certain exceptions.
-In the following example a test case for `productionFeature()` has been implemented. The test
-has been fully implemented, but the production code has not been implemented yet and throws an
-`UnsupportedOperationException` to indicate this. Once the production code has been implemented
-the test method will fail if the code were to return the wrong result and fail due to a failing
-assertion.
+In the following example a test case for `productionFeature()` has been implemented.
+The test has been fully implemented, but the production code has not been implemented yet and throws an `UnsupportedOperationException` to indicate this.
+Once the production code has been implemented the test method will fail if the code were to return the wrong result and fail due to a failing assertion.
```
Comment (Michael1993) [inline docs/expected-to-fail-tests.adoc:50]: Wait, is this correct? I thought it would always fail because the expectation is not fulfilled so it doesn't matter what's in the test.

Comment (knutwannheden) [reply docs/expected-to-fail-tests.adoc:50]: Maybe you can propose a better way to phrase this. By "Once the production code has been implemented" I mean that `productionFeature()` is now no longer throwing an `UnsupportedOperationException`, but instead maybe the wrong answer (e.g. `11`). Without `onExceptions` the test would continue to be skipped. With `onExceptions` the failing assertion will now cause the test to fail.

Perhaps it would make it clearer if I spell it out like that.

Comment (Michael1993) [reply docs/expected-to-fail-tests.adoc:50]:
```
Once the feature has been implemented `@ExpectedToFail` will fail the test as it no longer throws `UnsupportedOperationException`. If you don't use `onExceptions` in `@ExpectedToFail` you risk "masking" your implementation being faulty - which `@ExpectedToFail` would mark as passing.
```

What do you think?

I would also put a warning here to NOT use `onExceptions` for assertion - you can look at Nicolai's comment on the linked issue.

Comment (knutwannheden) [reply docs/expected-to-fail-tests.adoc]: I've rephrased the documentation yet yet again (and added a subsection header). I hope this will clarify things.

Please note that I am of course also happy with you changing the documentation to whatever you like. I am obviously not a technical writer 😄

**Review by Michael1993 — APPROVED** [review]
I'm happy with current state of the PR. I'd ask you to add yourself to the README as a contributor (but that's not mandatory, only if you want to). Thank you for your patience and hard work!

**Comment by knutwannheden** [comment]
> Thank you for your patience and hard work!

Thanks a lot for your thoroughness and patience!

**Comment by Bukama** [comment]
Running full run

File: src/main/java/org/junitpioneer/jupiter/ExpectedToFail.java [hunk 10]
```diff
@@ -33,6 +33,13 @@
  * of disabled tests over time.
  * </p>
  *
+ * <p>Further, the {@link #onExceptions()} attribute can be used to restrict the extension's behavior
+ * to specific exceptions. I.e. only if the test method end up throwing one of the specified exceptions
```
Comment (beatngu13) [inline src/main/java/org/junitpioneer/jupiter/ExpectedToFail.java:37]:
```suggestion
 * to specific exceptions. That is, only if the test method ends up throwing one of the specified exceptions
```

File: src/main/java/org/junitpioneer/jupiter/ExpectedToFailExtension.java [hunk 11]
```diff
@@ -42,7 +43,12 @@ private static void invokeAndInvertResult(Invocation<Void> invocation, Extension
 				throw t;
 			}
 
-			String message = getExpectedToFailAnnotation(extensionContext).value();
+			ExpectedToFail expectedToFail = getExpectedToFailAnnotation(extensionContext);
+			if (Stream.of(expectedToFail.onExceptions()).noneMatch(clazz -> clazz.isInstance(t))) {
+				fail("Test marked as 'expected to fail' failed with an unexpected exception", t);
```
Comment (beatngu13) [inline src/main/java/org/junitpioneer/jupiter/ExpectedToFailExtension.java:48]:
```suggestion
				fail("Test marked as temporarily 'expected to fail' failed with an unexpected exception", t);
```

**Review by nipafx — CHANGES_REQUESTED** [review]
Hey @knutwannheden, thank you for opening the PR - it looks really good. 😃  A few things could be changed - see below.

But, importantly, can you please open another PR [from a feature branch](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.adoc#branching-strategy) and not `main`?  Thanks!

**Comment by nipafx** [comment]
Oh, and if you rebase it on the current `main`, you can fix the merge conflict - it comes from the `README`, which now `.adoc` not `.md`.

**Comment by knutwannheden** [comment]
@nipafx Thanks for your review comments. I assume I should wait until the other maintainers have chimed in. In the meantime I've updated the PR with the latest changes from `main`.

**Comment by Bukama** [comment]
> @nipafx Thanks for your review comments. I assume I should wait until the other maintainers have chimed in. In the meantime I've updated the PR with the latest changes from `main`.

Well what @nipafx didn't wrote: We literally had a team meeting online stream about all Pioneer things yesterday :)

**Comment by nipafx** [comment]
I managed to push a small fix on the README order to your `main` branch, but it's a bit cumbersome. We can keep this PR as is, but please keep in mind for the next one, not to open it from your `main` branch.

Other than that, I will let you take a look at the two remaining comments I made above.

File: src/main/java/org/junitpioneer/jupiter/ExpectedToFail.java [hunk 12]
```diff
@@ -72,4 +78,9 @@
 	 */
 	String value() default "";
 
+	/**
+	 * Specifies which exceptions are expected to be thrown and will cause the test to be aborted rather than fail.
+	 */
+	Class<? extends Throwable>[] onExceptions() default { Throwable.class };
```
[the remark this answers is not in the document]
Comment (knutwannheden) [reply src/main/java/org/junitpioneer/jupiter/ExpectedToFail.java]: I renamed it to `withExceptions` now (notice the plural form). Only supporting a single exception type would however also work well for me.

File: docs/expected-to-fail-tests.adoc [hunk 13]
```diff
@@ -44,6 +44,25 @@ A custom message can be provided, explaining why the tested code is not working
 include::{demo}[tag=expected_to_fail_message]
 ----
 
+=== Only Abort on Expected Exceptions
+
+A test that is `@ExpectedToFail` will change its behavior by starting to actually fail, once the code under test behaves correctly.
+If the underlying failure changes, though, for example from an assertion error to an exception or from an `UnsupportedOperationException` of a formerly missing implementation to a runtime exception of buggy implementation, the `@ExpectedToFail`-test will keep passing.
+
+To better react to such changes, `@ExpectedToFail` has an attribute `onExceptions` that can be used to enumerate the exceptions which when thrown will result in an aborted (and thus passing) test.
```
Comment (Michael1993) [inline docs/expected-to-fail-tests.adoc:52]:
```suggestion
To better react to such changes, `@ExpectedToFail` has an attribute `withExceptions` that can be used to enumerate the exceptions which when thrown will result in an aborted (and thus passing) test.
```

File: src/main/java/org/junitpioneer/jupiter/ExpectedToFailExtension.java (hunk 11 above)
Comment (Michael1993) [reply src/main/java/org/junitpioneer/jupiter/ExpectedToFailExtension.java:48]: @knutwannheden You marked this as resolved but I don't see the change?

**Review by Michael1993 — CHANGES_REQUESTED** [review]
If you have the time, please write a few more tests to cover the lines marked in the Sonar analysis.

**Comment by knutwannheden** [comment]
@Michael1993 Sorry about the long delays!

File: src/main/java/org/junitpioneer/jupiter/ExpectedToFailExtension.java [hunk 14]
```diff
@@ -31,6 +32,11 @@ public void interceptTestMethod(Invocation<Void> invocation, ReflectiveInvocatio
 
 	private static void invokeAndInvertResult(Invocation<Void> invocation, ExtensionContext extensionContext)
 			throws Throwable {
+		ExpectedToFail expectedToFail = getExpectedToFailAnnotation(extensionContext);
+		if (expectedToFail.withExceptions().length == 0) {
+			fail("@ExpectedToFail withExceptions must not be empty");
```
Comment (Michael1993) [inline src/main/java/org/junitpioneer/jupiter/ExpectedToFailExtension.java:37]: This is not an `AssertionFailedError`, since it does not originate from the test itself - should be an `ExtensionConfigurationException`.

**Review by Michael1993 — APPROVED** [review]

File: docs/expected-to-fail-tests.adoc (hunk 9 above)
Comment (beatngu13) [reply docs/expected-to-fail-tests.adoc]: @Michael1993 given your approval on December 2, 2023, I mark this as resolved.

**Review by beatngu13 — APPROVED** [review]

**Comment by beatngu13** [comment]
Proposed commit message:

```
Add withExceptions to ExpectedToFail (#769 / #774)

Adds a new `withExceptions` attribute to `@ExpectedToFail`, which
allows to limit the scope of the extension to only consider the test
as successful if the thrown exception matches one of the given types.

Closes: #769
PR: #774
```

**Comment by beatngu13** [comment]
@knutwannheden thanks again! 🙏
